### PR TITLE
Add Edit Task feature implemented using MVVM

### PR DESCRIPTION
**Pull request by florina-muntenescu**

### CODE REVIEW

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 1]
```diff
@@ -30,23 +28,38 @@
...
 
+    @Nullable
+    private CompositeSubscription mSubscription;
+
+    @Nullable
+    private AddEditTaskViewModel mViewModel;
+
     public static AddEditTaskFragment newInstance() {
         return new AddEditTaskFragment();
     }
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:64]: The thing about next line `public AddEditTaskFragment()` as I cannot put the comment there. Do you really need to explicitly create that ctor?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 2]
```diff
@@ -58,18 +71,36 @@ public AddEditTaskFragment() {
...
         super.onResume();
-        mPresenter.subscribe();
+        bind();
     }
 
     @Override
     public void onPause() {
         super.onPause();
-        mPresenter.unsubscribe();
+        unbind();
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: Move before onPause

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 3]
```diff
@@ -30,23 +28,38 @@
...
-    private AddEditTaskContract.Presenter mPresenter;
-
+    @Nullable
     private TextView mTitle;
 
+    @Nullable
     private TextView mDescription;
 
+    @Nullable
     private String mEditedTaskId;
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: Should this be in the vm, looks like logic

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 4]
```diff
@@ -78,30 +109,66 @@ public void onActivityCreated(Bundle savedInstanceState) {
...
                 (FloatingActionButton) getActivity().findViewById(R.id.fab_edit_task_done);
         fab.setImageResource(R.drawable.ic_done);
         fab.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
                 if (isNewTask()) {
-                    mPresenter.createTask(
-                            mTitle.getText().toString(),
-                            mDescription.getText().toString());
+                    createTask();
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: I would do it differently.
The better approach would be to have a method that returns Observable of the click events and use it in bind.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 5]
```diff
@@ -78,30 +109,66 @@ public void onActivityCreated(Bundle savedInstanceState) {
...
             }
         });
     }
 
+    private void updateTask() {
+        Preconditions.checkNotNull(mTitle);
+        Preconditions.checkNotNull(mDescription);
+
+        getSubscription().add(getViewModel()
+                .updateTask(mTitle.getText().toString(), mDescription.getText().toString())
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: Add schedulers

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 6]
```diff
@@ -146,4 +212,16 @@ private void setTaskIdIfAny() {
     private boolean isNewTask() {
         return mEditedTaskId == null;
     }
+
+    @NonNull
+    private AddEditTaskViewModel getViewModel() {
+        Preconditions.checkNotNull(mViewModel);
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: `return Preconditions.checkNotNull(mViewModel);`

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 7]
```diff
@@ -146,4 +212,16 @@ private void setTaskIdIfAny() {
...
+        return mViewModel;
+    }
+
+    @NonNull
+    private CompositeSubscription getSubscription() {
+        Preconditions.checkNotNull(mSubscription);
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: `return Preconditions.checkNotNull(mSubscription);`

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 8]
```diff
@@ -111,29 +178,28 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
         return root;
     }
 
-    @Override
-    public void showEmptyTaskError() {
+    private void showEmptyTaskError() {
         Snackbar.make(mTitle, getString(R.string.empty_task_message), Snackbar.LENGTH_LONG).show();
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:159]: add `Preconditions.checkNotNull(mTitle);`

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+ * Listens to user actions from the UI ({@link AddEditTaskFragment}), retrieves the task from
+ * repository, creates and updates a task.
+ */
+public class AddEditTaskViewModel {
+
+    @NonNull
+    private final TasksDataSource mTasksRepository;
+
+    @Nullable
+    private String mTaskId;
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: Can be final

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Creates and saves a new task with a title and description.
+     *
+     * @param title       the title of the new task
+     * @param description the description of the new task.
+     * @return a stream that emits only once, after the task was created. The stream will emit an
+     * error if the task is empty.
+     */
+    @NonNull
+    public Observable<Void> createTask(@Nullable String title,
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: You can use `Completable` instead

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+     * Updates the title and the description of the the task with the current task id in the
+     * repository.
+     *
+     * @param title       the new title of the task
+     * @param description the new description of the task
+     * @return a stream that emits only once, after the task was updated. The stream will emit an
+     * error if there is no task id or if the task is empty.
+     */
+    @NonNull
+    public Observable<Void> updateTask(@Nullable String title,
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: You can use `Completable` instead

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+                                       @Nullable String description) {
+        if (mTaskId == null) {
+            return Observable.error(new Exception("updateTask() was called but task is new."));
+        }
+        return saveTask(new Task(title, description, mTaskId));
+    }
+
+    @NonNull
+    private Observable<Void> saveTask(@NonNull Task newTask) {
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: You can use `Completable` instead

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        if (newTask.isEmpty()) {
+            return Observable.error(new Exception("Trying to save an empty task"));
+        }
+
+        return Observable.just(newTask)
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: If you would be using `Completable`, then do `Completable.fromCallable(() -> mTasksRepository.saveTask(task))` instead

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        mTestSubscriber.assertValueCount(1);
+    }
+
+    @Test
+    public void createTask_doesNotSaveTask_whenTaskEmpty() {
+        // Given no task id set
+        mViewModel = new AddEditTaskViewModel(null, mTasksRepository);
+
+        // When creating a task with empty title and description
+        mViewModel.createTask(null, null).subscribe(mTestSubscriber);
```
Comment (tomaszpolanski) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java]: Use empty `subscribe`

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        mTestSubscriber.assertValueCount(1);
+    }
+
+    @Test
+    public void updateTask_doesNotSaveTask_whenTaskEmpty() {
+        // Given a view model with taskId set
+        mViewModel = new AddEditTaskViewModel(TASK.getId(), mTasksRepository);
+
+        // When updating a task with empty title and description
+        mViewModel.updateTask(null, null).subscribe(mTestSubscriber);
```
Comment (tomaszpolanski) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java]: Use empty `subscribe`

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java (hunk 10 above)
Comment (florina-muntenescu) [reply todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: Never head of `Completable` before. I'll check it out.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 16]
```diff
@@ -30,78 +28,117 @@
...
-                    mPresenter.createTask(
-                            mTitle.getText().toString(),
-                            mDescription.getText().toString());
-                } else {
-                    mPresenter.updateTask(
-                            mTitle.getText().toString(),
-                            mDescription.getText().toString());
-                }
-
+                saveTask();
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:113]: Similar question as in other PR, if you want to have all reactive thing in bind, you can make an observable from click events. But is up to you

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 17]
```diff
@@ -30,78 +28,117 @@
...
             }
         });
     }
 
+    private void saveTask() {
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:118]: If the click would be an observable, you would be able to put it into bind

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+import android.support.annotation.Nullable;
+
+import com.example.android.architecture.blueprints.todoapp.data.Task;
+import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource;
+
+import java.util.concurrent.Callable;
+
+import rx.Completable;
+import rx.Observable;
+import rx.functions.Action0;
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java:11]: Clean up imports

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+     * @param taskId          ID of the task to edit or null for a new task
+     * @param tasksRepository a repository of data for tasks
+     */
+    public AddEditTaskViewModel(@Nullable String taskId,
+                                @NonNull TasksDataSource tasksRepository) {
+        mTaskId = taskId;
+        mTasksRepository = checkNotNull(tasksRepository, "tasksRepository cannot be null!");
+    }
+
+    @NonNull
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: NonNull not needed

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    private void saveTask(@NonNull Task newTask) throws Exception {
+        if (newTask.isEmpty()) {
+            throw new Exception("Trying to save an empty task");
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: Exception is not the pretty way of doing this, use maybe `RuntimeException`.
Also, you will be able to remove `throws Exception`

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Creates or updates a task.
+     *
+     * @param title       the title of the task
+     * @param description the description of the task
+     * @return a stream that emits only once, after the task was created. The stream will emit an
+     * error if the task is empty.
+     */
+    @NonNull
+    public Completable saveTask(@Nullable String title,
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: If you remove the `Exception` from other `saveTask` you would be able to rewrite it like this:

```
    @NonNull
    public Completable saveTask(@Nullable final String title,
                                @Nullable final String description) {
        return Completable.fromAction(new Action0() {
            @Override
            public void call() {
                Task newTask = mTaskId == null
                        ? new Task(title, description)
                        : new Task(title, description, mTaskId);
                saveTask(newTask);
            }
        });
    }
```

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Unit tests for the implementation of {@link AddEditTaskViewModel}.
+ */
+public class AddEditTaskViewModelTest {
+
+    private final String TITLE = "title";
```
Comment (tomaszpolanski) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java]: Make those final, and maybe rename to lower case

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    @Test
+    public void saveTask_emitsError_whenTaskEmpty() {
+        // Given no task id set
+        mViewModel = new AddEditTaskViewModel(null, mTasksRepository);
+
+        // When saving a task with empty title and description
+        mViewModel.saveTask(null, null).subscribe(mTestSubscriber);
+
+        // An error is emitted
+        mTestSubscriber.assertError(Exception.class);
```
Comment (tomaszpolanski) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModelTest.java]: If you choose to use `RuntimeException`, change it here

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java (hunk 20 above)
Comment (florina-muntenescu) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java]: If I use RuntimeException it crashes the app.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 24]
```diff
@@ -30,78 +28,123 @@
...
+    private void unbindViewModel() {
+        getSubscription().unsubscribe();
     }
 
     @Override
     public void onActivityCreated(Bundle savedInstanceState) {
         super.onActivityCreated(savedInstanceState);
 
-        setTaskIdIfAny();
+        mViewModel = Injection.provideAddEditTaskViewModel(getTaskId(), getContext());
```
Comment (tomaszpolanski) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java:105]: Won't compile. Invert arguments

**Pull request closed without merging**